### ISSUE TITLE
Thin LTO + `-Os` gives vmlinux.o of almost the same size as thin LTO + `-O2`

### ISSUE THREAD

**Issue by emojifreak** [ClangBuiltLinux/linux#1559]
https://github.com/ClangBuiltLinux/linux/issues/1550#issuecomment-1001118975 gave `vmlinux.o` with 5 GB, which was produced with thin LTO + `-O2`. Changing `-O2` to `-Os` gives

```
$ LANG=C ls -ltr *vmlin* .*vmlinu*
-rw-r--r-- 1 ryutaroh ryutaroh 5052621640 Dec 27 16:45 vmlinux.o
-rw-r--r-- 1 ryutaroh ryutaroh    1543609 Dec 27 16:45 vmlinux.symvers
-rw-r--r-- 1 ryutaroh ryutaroh 1813722834 Dec 27 16:47 .tmp_vmlinux.kallsyms1.map
```

A reproducing script and `.config` producing the above result are attached.

This is in sharp contrast with x86. `-Os` gives significantly smaller `vmlinux.o` and suppressed https://github.com/ClangBuiltLinux/linux/issues/1556.
[config.txt.gz](https://github.com/ClangBuiltLinux/linux/files/7782300/config.txt.gz)
[clang-build-arch30s.sh.gz](https://github.com/ClangBuiltLinux/linux/files/7782301/clang-build-arch30s.sh.gz)

**Comment by emojifreak** [comment]
The title and the original report were a bit incorrect. The same phenomenon appears with `x86`.

**Comment by nathanchance** [comment]
Thinking more about this, I am not really sure there is much of a bug here. The difference between `-O2` and `-Os` according to [GCC's documentation](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html):

```
-Os

Optimize for size. -Os enables all -O2 optimizations except those that often increase code size:

-falign-functions  -falign-jumps 
-falign-labels  -falign-loops 
-fprefetch-loop-arrays  -freorder-blocks-algorithm=stc
```

which is not much. Let's take LTO out of the equation entirely and see what kind of difference the optimization level makes with just a plain `allyesconfig` build. Writing and running [a small script](https://raw.githubusercontent.com/nathanchance/repro-scripts/fbbfa5b7ccbc472cf32d6c32d4d158039849f3c4/diff-arm64-x86_64-allyesconfig-opt.fish) to do all the builds and diff the results shows:

```
ClangBuiltLinux clang version 14.0.0 (https://github.com/llvm/llvm-project 09949c48cb18f70d805caffd6bcfc53aaae7f040)
gcc (GCC) 11.2.1 20220109
aarch64-linux-gnu-gcc (GCC) 11.2.0

ARCH=arm64 CC=clang:
  -O2: 645.96 MB (645,955,584 bytes)
  -Os: 624.52 MB (624,517,120 bytes)
  Difference: -21.44 MB (-3.43%)

ARCH=arm64 CC=gcc:
  -O2: 808.10 MB (808,099,840 bytes)
  -Os: 772.80 MB (772,804,608 bytes)
  Difference: -35.3 MB (-4.57%)

ARCH=x86_64 CC=clang:
  -O2: 1.40 GB (1,396,158,464 bytes)
  -Os: 1.34 GB (1,336,377,344 bytes)
  Difference: -59.78 MB (-4.47%)

ARCH=x86_64 CC=gcc:
  -O2: 1.48 GB (1,484,980,224 bytes)
  -Os: 1.32 GB (1,324,093,440 bytes)
  Difference: -160.89 MB (-12.15%)
```

In the best case scenario, there is only a savings of 12% but it seems like for most cases, it is going to be less than 5%, which seems to line up with the observations of the original post.

Long story short, I don't think there is anything for us to fix here. #1550 and #1556 should still probably be looked into at some point but expecting `-Os` to solve the problem of making the kernel image too large due to building all drivers into the kernel is unrealistic from my point of view. You should either use `-Os` in combination with building what you need OR you are going to use `-O2` and build a fully modular kernel.

**Comment by emojifreak** [comment]
Thank you very much for your investigation. Let us close this issue.
